topi/python/topi/cuda/rcnn/proposal.py

```diff
@@ -0,0 +1,355 @@
+        2-D with shape [batch, 3]
+
+    out_buf : tvm.schedule.Buffer
+        3-D with shape [batch, num_bbox, 5]
+        The last dimension is in format of [w_start, h_start, w_end, h_end, score]
+
+    scales : list/tuple of float
+        Scales of anchor windoes.
+
+    ratios : list/tuple of float
+        Ratios of anchor windoes.
+
+    feature_stride : int
+        The size of the receptive field each unit in the convolution layer of the rpn, for example
+        the product of all stride's prior to this layer.
+
+    rpn_min_size : int
+        Minimum height or width in proposal.
+
+    iou_loss : bool
+        Usage of IoU loss.
+
+    Returns
+    -------
+    stmt : Stmt
+        The result IR statement.
+    """
+    batch, num_anchors, height, width = get_const_tuple(cls_prob_buf.shape)
+    num_anchors //= 2
+    max_threads = int(tvm.target.current_target(allow_none=False).max_num_threads)
+    nthread_tx = max_threads
+    nthread_bx = (batch * height * width) // max_threads + 1
+    tx = tvm.thread_axis("threadIdx.x")
+    bx = tvm.thread_axis("blockIdx.x")
+    tid = bx * max_threads + tx
+    ib = tvm.ir_builder.create()
+    ib.scope_attr(tx, "thread_extent", nthread_tx)
+    ib.scope_attr(bx, "thread_extent", nthread_bx)
+
+    p_score = ib.buffer_ptr(cls_prob_buf)
+    p_delta = ib.buffer_ptr(bbox_pred_buf)
+    p_im_info = ib.buffer_ptr(im_info_buf)
+    p_out = ib.buffer_ptr(out_buf)
+
+    with ib.if_scope(tid < batch * height * width):
+        w = tid % width
+        h = (tid // width) % height
+        b = tid // width // height
+
+        for k in range(num_anchors):
+            out_index = tid * num_anchors + k
+            ratio = ratios[k // len(scales)]
+            scale = scales[k % len(scales)]
+            anchor = generate_anchor(ratio, scale, feature_stride)
+            im_height = p_im_info[b * 3]
+            im_width = p_im_info[b * 3 + 1]
+            x1 = anchor[0] + w * feature_stride
+            y1 = anchor[1] + h * feature_stride
+            x2 = anchor[2] + w * feature_stride
+            y2 = anchor[3] + h * feature_stride
+
+            delta = [p_delta[((((b * num_anchors + k) * 4 + i) * height + h) * width + w)]
+                     for i in range(4)]
+            regression_func = reg_iou if iou_loss else reg_bbox
+            pred_x1, pred_y1, pred_x2, pred_y2 = regression_func(x1, y1, x2, y2, *delta)
+
+            pred_x1 = tvm.max(tvm.min(pred_x1, im_width - 1.0), 0.0)
+            pred_y1 = tvm.max(tvm.min(pred_y1, im_height - 1.0), 0.0)
+            pred_x2 = tvm.max(tvm.min(pred_x2, im_width - 1.0), 0.0)
+            pred_y2 = tvm.max(tvm.min(pred_y2, im_height - 1.0), 0.0)
+
+            real_height = (im_height / feature_stride).astype('int32')
+            real_width = (im_width / feature_stride).astype('int32')
+
+            bbox_w = pred_x2 - pred_x1 + 1.0
+            bbox_h = pred_y2 - pred_y1 + 1.0
+            min_size = p_im_info[b * 3 + 2] * rpn_min_size
```

Comment:
what is the third dim of im_info?

Reply:
the third dim is the scale of the origin image size to the fixed input size of the network